core/src/replay_stage.rs

```diff
@@ -1685,21 +1707,24 @@ impl ReplayStage {
     fn replay_blockstore_into_bank(
         bank: &Arc<Bank>,
         blockstore: &Blockstore,
-        bank_progress: &mut ForkProgress,
+        replay_stats: &RwLock<ReplaySlotStats>,
+        replay_progress: &RwLock<ConfirmationProgress>,
         transaction_status_sender: Option<&TransactionStatusSender>,
         replay_vote_sender: &ReplayVoteSender,
         transaction_cost_metrics_sender: Option<&TransactionCostMetricsSender>,
         verify_recyclers: &VerifyRecyclers,
     ) -> result::Result<usize, BlockstoreProcessorError> {
-        let tx_count_before = bank_progress.replay_progress.num_txs;
+        let mut w_replay_stats = replay_stats.write().unwrap();
```

Comment:
does holding the lock this long essentially single thread what im guessing is the longest part of replay, the execution?

Reply:
You're right that this is expected to be the longest part of replay. However, this shouldn't lock out the other forks because these stats are local to each fork in the progress map. I believe getting entries out of the fork progress map (or inserting them) is the major point of lock contention

Reply:
ah okay, nice!

Reply:
> these stats are local to each fork in the progress map

Why do we need `RwLock` then here?
It is still not clear to me how this does not cause lock contention.

Reply:
I just confirmed that we can grab a write lock for `replay_stats` and `replay_progress` for two different slots at the same time.

Here's my understanding of why it's necessary and how it works (if there's a better way of doing this where we can eliminate the lock, even better!):
1. We create a lock around `ProgressMap` because different threads may need to write this to insert a `ForkProgress` entry
2. We grab a mutable reference to the individual `ForkProgress` entry associated with the current slot we're working on (e.g. Slot X)
3. We grab a handle to `replay_stats` and `replay_progress` locks owned by `ForkProgress`, which are associated w/ Slot X
4. We drop the overarching lock for `ProgressMap` <-- At this point, separate forks can start replaying in parallel (slots besides Slot X)
5. We acquire write lock for `replay_stats` and `replay_progress` associated w/ Slot X